src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+            // `future_rs.as_mut().poll(...)`, so continue looping
+        }
+    }
+}
+
+pub(crate) fn iter_result(result: IterNextOutput<PyObject, PyObject>) -> PyResult<PyObject> {
+    match result {
+        IterNextOutput::Yield(ob) => Ok(ob),
+        IterNextOutput::Return(ob) => Err(PyStopIteration::new_err(ob)),
+    }
+}
+
+#[pymethods(crate = "crate")]
+impl Coroutine {
+    fn send(&mut self, py: Python<'_>, _value: &PyAny) -> PyResult<PyObject> {
+        iter_result(self.poll(py, None)?)
+    }
+
+    fn throw(&mut self, py: Python<'_>, exc: &PyAny) -> PyResult<PyObject> {
+        iter_result(self.poll(py, Some(exc))?)
+    }
+
+    fn close(&mut self) {
+        self.future.take();
+    }
+
+    fn __await__(self_: Py<Self>) -> Py<Self> {
+        self_
+    }
+
+    fn __next__(&mut self, py: Python<'_>) -> PyResult<IterNextOutput<PyObject, PyObject>> {
+        self.poll(py, None)
+    }
+}
+
+/// `asyncio.Future` wrapper, implementing [`ArcWake`] by calling `Future.set_result`.
+///
+/// It embeds the event loop thread id to determine if it's necessary to use `call_soon_threadsafe`.
+struct PyFutureWaker {
+    future: PyObject,
+    thread_id: ThreadId,
+}
+
+impl PyFutureWaker {
+    fn wake_same_thread(&self, py: Python<'_>) {
+        // `set_result` can be called directly because in the event loop thread
+        self.future
+            .call_method1(py, intern!(py, "set_result"), (py.None(),))
+            .expect("error while calling Future.set_result");
+    }
+
+    fn wake_threadsafe(&self, py: Python<'_>) {
+        // `set_result` must be called in event loop thread, so it requires `call_soon_threadsafe`
+        let set_result = self
+            .future
+            .getattr(py, intern!(py, "set_result"))
+            .expect("error while calling Future.set_result");
+        let event_loop = self
+            .future
+            .call_method0(py, intern!(py, "get_loop"))
+            .expect("error while calling Future.get_loop");
+        if let err @ Err(_) = event_loop.call_method1(
+            py,
+            intern!(py, "call_soon_threadsafe"),
+            (set_result, py.None()),
+        ) {
+            // `call_soon_threadsafe` will raise if the event loop is closed;
+            // instead of catching an unspecific `RuntimeError`, check directly if it's closed.
+            let is_closed = event_loop
+                .call_method0(py, "is_closed")
+                .and_then(|res| res.extract(py))
+                .unwrap_or(false);
+            // if the loop is not closed, panic
+            if !is_closed {
+                err.expect("error while calling EventLoop call_soon_threadsafe");
+            }
+        }
+    }
+}
+
+impl ArcWake for PyFutureWaker {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        // Choose appropriate wake method depending on wether it is called on the same thread
+        // than the event loop
+        if current_thread_id() == arc_self.thread_id {
```

Comment:
Just to make sure because I do not know this and it seems to be assumed here: The asyncio event loop can nether be moved to a different thread? Not even stopped on one and restart on another?

Reply:
That's a good concern, and I think it's indeed possible to move the event loop to a different thread. So this "optimization" is wrong, I will remove it.